### ISSUE TITLE
Should decide best way to test fundamental JitBuilder services

### ISSUE THREAD

**Issue by mstoodle** [eclipse/omr#627]
Services like "LessThan" and its variants translate into extremely straight-forward Testarossa IL. There are many variants, especially once you take the data type into account, and they all basically boil down to an opcode.

I'm not suggesting we should not have tests for these basic services, but given all the variations it's currently pretty painful to write test for, and at least as error prone to create the test as to write the service being tested.

I'm wondering if there's a better way to do this testing?

We have also been putting tests for JitBuilder services into the jitbuilder/release/src directory. That's a fine location because then they can also serve as examples for people who want to see how various JitBuilder services can be used, but if we start creating tests with tons of variations I have a feeling those tests will not be very useful as examples.

Maybe we should return to the notion to add testing into the fvtest/compilertest/ directory along the lines of the BuilderTest.cpp file (which is run as part of our make test). These tests do not use JitBuilder per se, but they test the underlying services that JitBuilder exposes.

I am currently leaning towards implementing/moving the basic testing of the "*Builder" classes into the fvtest/compilertest directory and treating jitbuilder/release/src as a "repository" of nice examples for people to emulate.

Does the rest of the community agree with that approach? Any other suggestions for the builder API testing that I haven't thought of?

**Comment by charliegracie** [comment]
I like that approach and I am in favor of the project moving in that direction

**Comment by Leonardo2718** [comment]
I agree that we should be putting more of our testing in the `fvtest/compilertest/` directory.

To address the issue of testing different variations of a service, I think it should be possible to develop a lightweight framework that can take care of most of the boiler plate code. My understanding is that most of the heavy work due to the large amount of boiler plate that's required. In theory a combination of mocking and data-driven testing would be the right answer here. Realistically, I think we need to focus on reducing the amount of boiler plate code while minimizing the side-effects this will have on test cases.
